### DISCUSSION TITLE
nm does not work

### DISCUSSION THREAD

**Discussion by ballapete** [mesonbuild/meson#11131] [Q&A]
Hello!

On old Mac OS X systems Leopard (10.5.8) and Tiger (10.4.11) the nm command seems to have a different syntax than on more recent versions. When building for example libepoxy meson (0.58.2 is recorded in the log) complains:

```
WARNING: ['nm'] does not work. Relinking will always happen on source changes.
error: nm: invalid argument --
Usage: nm [-agnopruUmxjlfAP[s segname sectname] [-] [-t format] [[-arch <arch_flag>] ...] [file ...]
```

How can I make meson show which nm command it is invoking? Where in its source code are the arguments for nm written down?

--

Pete

**Comment by eli-schwartz** [comment]
See https://github.com/mesonbuild/meson/blob/9074ad93c6675cdb603b85d85a8090100058f81b/mesonbuild/scripts/symbolextractor.py#L135-L154

If it's possible to detect both cases that would be nice.

**Comment by ballapete** [comment]
First answer: Mac OS X 10.4, Tiger, has old nm, Mac OS X 10.5, Leopard, certainly as well (exact data in a few days). MacOS High Sierra, Version 10.13 or Darwin 17, and macOS Monterey, Version 12 or Darwin 21, use the modern nm.

I'll try to find out more. Another option would be to prepare a patch that on the oldish OS versions meson invokes nm with **-g** – _Display only global (external) symbols_ – or **-U** – _Don't display undefined symbols_ –. I am using the MacPorts package manager, and the mentioned patch could just stay there and meson untouched?

Another option could be to check whether **llvm-nm** exists (in /opt/local/libexec/llvm-*/bin, I think the sym-link /opt/local/bin/llvm-nm is also usually provided). This tool might be already standard for the MacPorts users that have installed Clang and llvm. The latter "port" provides llvm-nm since Tiger.

--

Pete

**Comment by ballapete** [comment]
This patch should work:

```
--- mesonbuild/scripts/symbolextractor.py~	1970-01-01 01:00:00.000000000 +0100
+++ mesonbuild/scripts/symbolextractor.py	2022-12-01 18:32:05.000000000 +0100
@@ -146,8 +146,7 @@
             break
     result = [arr[match + 2], arr[match + 5]] # Libreoffice stores all 5 lines but the others seem irrelevant.
     # Get a list of all symbols exported
-    output = call_tool('nm', ['--extern-only', '--defined-only',
-                              '--format=posix', libfilename])
+    output = call_tool('nm', ['-a', '-U', '-P', libfilename])
     if not output:
         dummy_syms(outfilename)
         return
```

nm first outputs an empty line and then the pathname of the library. I therefore added two empty lines in the output of llvm-nm to easier compare the two outputs:

```
pete 253 /\ llvm-nm --extern-only /opt/local/lib/libcharset.1.dylib	pete 254 /\ nm -a /opt/local/lib/libcharset.1.dylib
																    
									/opt/local/lib/libcharset.1.dylib(single module):	    
00000000 U ____mb_cur_max_l						         U ____mb_cur_max_l				    
00000b9c t ___initialize_Cplusplus					00000b9c t ___initialize_Cplusplus			    
00000b7c t __dyld_func_lookup						00000b7c t __dyld_func_lookup				    
00000000 T __mh_dylib_header						00000000 t __mh_dylib_header				    
00000da6 s _alias_table							00000da6 s _alias_table					    
00000ce4 T _libcharset_set_relocation_prefix				00000ce4 T _libcharset_set_relocation_prefix		    
00000bf8 T _locale_charset						00000bf8 T _locale_charset				    
00000000 U _nl_langinfo							         U _nl_langinfo					    
00000000 U _strcmp							         U _strcmp					    
00000000 U _uselocale							         U _uselocale					    
00000b74 t cfm_stub_binding_helper					00000b74 t cfm_stub_binding_helper			    
0000101c s dyld__mach_header						0000101c d dyld__mach_header				    
00000b44 t dyld_stub_binding_helper					00000b44 t dyld_stub_binding_helper			    

pete 256 /\ llvm-nm --defined-only /opt/local/lib/libcharset.1.dylib	pete 257 /\ nm -U /opt/local/lib/libcharset.1.dylib
															   
									/opt/local/lib/libcharset.1.dylib(single module):  
00000b9c t ___initialize_Cplusplus					00000b9c t ___initialize_Cplusplus		   
00000b7c t __dyld_func_lookup						00000b7c t __dyld_func_lookup			   
00000000 T __mh_dylib_header						00000000 t __mh_dylib_header			   
00000da6 s _alias_table							00000da6 s _alias_table				   
00000ce4 T _libcharset_set_relocation_prefix				00000ce4 T _libcharset_set_relocation_prefix	   
00000bf8 T _locale_charset						00000bf8 T _locale_charset			   
00000b74 t cfm_stub_binding_helper					00000b74 t cfm_stub_binding_helper		   
0000101c s dyld__mach_header						0000101c d dyld__mach_header			   
00000b44 t dyld_stub_binding_helper					00000b44 t dyld_stub_binding_helper		   

pete 258 /\ llvm-nm --format=posix /opt/local/lib/libcharset.1.dylib	pete 259 /\ nm -P /opt/local/lib/libcharset.1.dylib 
															    
									/opt/local/lib/libcharset.1.dylib(single module):   
____mb_cur_max_l U 00000000						____mb_cur_max_l U 0 0				    
___initialize_Cplusplus t 00000b9c					___initialize_Cplusplus t b9c 0			    
__dyld_func_lookup t 00000b7c						__dyld_func_lookup t b7c 0			    
__mh_dylib_header T 00000000						__mh_dylib_header t 0 0				    
_alias_table s 00000da6							_alias_table s da6 0				    
_libcharset_set_relocation_prefix T 00000ce4				_libcharset_set_relocation_prefix T ce4 0	    
_locale_charset T 00000bf8						_locale_charset T bf8 0				    
_nl_langinfo U 00000000							_nl_langinfo U 0 0				    
_strcmp U 00000000							_strcmp U 0 0					    
_uselocale U 00000000							_uselocale U 0 0				    
cfm_stub_binding_helper t 00000b74					cfm_stub_binding_helper t b74 0			    
dyld__mach_header s 0000101c						dyld__mach_header d 101c 0			    
dyld_stub_binding_helper t 00000b44					dyld_stub_binding_helper t b44 0	
```	    

BTW, the original file's date is incorrectly extracted by the programme MacPorts uses for this task. Actually this fault happens only with Python related archives… (that are not packed by gtar?)

**Comment by ballapete** [comment]
The patch works on PPC Mac OS X 10.4.11, Tiger, for meson 0.58.2. I could build dav1d. Before I saw a complaint about nm, after it was gone.

I found that on macOS High Sierra, Version 10.13 or Darwin 17, and macOS Monterey, Version 12 or Darwin 21, /usr/bin/nm understands both, -a **and** --extern-only.

So -a/-U/-P could be used on all Macs?

--

Pete

**Comment by ballapete** [comment]
On PPC Mac OS X 10.5.8, Leopard, or Darwin 9, /usr/bin/nm understands only the BSD syntax (and it actually has a bug, -U is not accepted, although documented as working accordingly to Tiger's nm). llvm-nm from MacPorts works here, and also an own nm-classic which is provided by cctools. And, as the name suggests, understands only short BSD syntax. For ease of use a symbolic link /opt/local/bin/nm is created that points to /opt/local/bin/nm-classic.

llvm-nm is provided by llvm-3.4 on both, Tiger and Leopard. This is the latest llvm version that built on PowerPC hardware. The build process seems to need Python 2.7, an outdated package. llvm does not seem to play a role on both Tiger and Leopard anymore, its job is performed by GCC 7, an effort is going on to bring GCC 12 to these very old Macs.

**Comment by ballapete** [comment]
I patched meson 0.63.3 on macOS Monterey, Version 12.6, and macOS High Sierra, Version 10.13.6. On both Macs dav1d 1.0.0 and libepoxy 1.5.0 built without making nm complain.

**Comment by TurtleWilly** [comment]
2024, OS X 10.10.5 here with XCode 7.2. No idea what version `nm` is (it fails with `--version` 😎 ) and I encounter the very same problem. I applied above patch and things work fine. Can we get this fix applied? Upvote from me!

**Comment by rgommers** [comment]
I have opened a PR to change to the one-character form: https://github.com/mesonbuild/meson/pull/13336.

The discussion above is largely consistent, except for `-a` and/or `-g` used to replace `--extern-only`. `-g` seems correct, that is the one-character form of `--extern-only` (see, e.g. the [GNU nm man page](https://linux.die.net/man/1/nm)). `-a` does something quite different.
